# Midterm (3rd Year IT): Coding + LLM Critique
Transparency with ChatGPT Use, Code Review, and Improvement

# Problems (Do All 10)
___

# Task #1. String Normalizer & Finder. 
Write a function normalize_and_find(text:
str, needle: str) -> int that: (1) trims whitespace on both ends, (2) converts to lowercase, then (3) returns the index of the first occurrence of needle using
a safe search (return −1 if not found). Discuss alternatives (find vs. index), and
how your choice affects error handling. Anchor concepts: stripping, case conversion,
find/replace, string immutability.


# Prompt (paste verbatim)
Write a Python function normalize_and_find(text, needle) that trims whitespace on both ends, converts both strings to lowercase, and returns the index of the first occurrence of needle in text. If not found, return -1. Discuss the difference between find() and index(), and how this choice affects error handling. Add a quick test.


In [1]:
def normalize_and_find(text, needle):
    # Step 1: Trim whitespace and convert to lowercase
    text = text.strip().lower()
    needle = needle.strip().lower()

    # Step 2: Find the first occurrence of needle in text
    index = text.find(needle)

    # Step 3: Return the index (or -1 if not found)
    return index


# 🔍 Quick test
print(normalize_and_find("   Hello World   ", "WORLD"))  # Output: 6
print(normalize_and_find("  Python Exam  ", "test"))     # Output: -1


6
-1


# Critique

# Correctness: 
Correctly normalizes both strings using .strip() and .lower(), then finds the first occurrence with .find(). The method returns -1 when not found, matching the spec. Behavior is correct for given tests.
# Complexity: 
O(n + m) time where n is length of text and m is length of needle. O(n + m) space due to creating new normalized strings. This is optimal for the string find operation.
# Robustness:  
Works for basic cases but lacks edge case handling. According to string basics, should consider empty strings and boundary conditions. No type validation is present.
# Readability: 
Step-by-step comments are clear but verbose for simple string operations. The intermediate index variable adds no value—directly returning the result would be cleaner. Emoji in comments is non-standard.                    Following string method conventions from Lesson One, the code could be more concise while remaining clear.
# Faithfulness: 
Uses core string methods from Lesson One (.strip(), .lower(), .find()) correctly. These are fundamental string manipulation operations covered in the strings lesson. However, the function mutates the        parameter names by reassigning, which can be confusing.

Reference: Lesson One - Strings (string methods: .strip(), .lower(), .find())

 # Improved Code

In [2]:
def normalize_and_find(text: str, needle: str) -> int:
    """
    Find the first occurrence of needle in text after normalization.
    
    Both strings are trimmed of whitespace and converted to lowercase
    before searching.
    
    Returns the index of first occurrence, or -1 if not found.
    """
    normalized_text = text.strip().lower()
    normalized_needle = needle.strip().lower()
    return normalized_text.find(normalized_needle)


if __name__ == "__main__":
    # Basic functionality
    assert normalize_and_find("   Hello World   ", "WORLD") == 6
    assert normalize_and_find("  Python Exam  ", "test") == -1
    
    # Edge cases from string basics
    assert normalize_and_find("", "") == 0  # Empty needle found at index 0
    assert normalize_and_find("abc", "") == 0
    assert normalize_and_find("", "x") == -1
    assert normalize_and_find("test", "test") == 0
    assert normalize_and_find("  TEST  ", "test") == 0
    
    print("All tests passed.")

All tests passed.


# Task #2. Leetspeak Translator (Parametric). 
Generalize our leetspeak idea: write translate(text:
str, mapping: dict[str,str]) -> str that applies a mapping (e.g., {’a’:’4’,’e’:’3’,...})
over lowercase letters only, leaving others unchanged. Explain why strings are immutable and show an efficient approach (e.g., list join vs. repeated concatenation).
Anchor concepts: replace, iteration strategies, efficiency. Reference: :contentReference[oaicite:7]index=7


# Prompt (paste verbatim)
Write a Python function translate(text, mapping) that applies a given mapping (e.g., {'a': '4', 'e': '3', ...}) to lowercase letters only, leaving all other characters unchanged. Explain why strings in Python are immutable and show an efficient approach for building the result (e.g., using list join instead of repeated string concatenation). Add a quick test.

In [4]:
def translate(text, mapping):
    # Efficiently build result using a list
    result_chars = []

    for ch in text:
        # Apply mapping only to lowercase letters
        if ch in mapping:
            result_chars.append(mapping[ch])
        else:
            result_chars.append(ch)

    # Join the list into a final string
    return ''.join(result_chars)


# 🔍 Quick test
leet_map = {'a': '4', 'e': '3', 'i': '1', 'o': '0'}
print(translate("Hello, World!", leet_map))  # Output: "H3ll0, W0rld!"
print(translate("Python is awesome", leet_map))  # Output: "Pyth0n 1s 4w3s0m3"

H3ll0, W0rld!
Pyth0n 1s 4w3s0m3


# Critique

# Correctness: 
The function correctly applies the mapping to characters found in the dictionary and leaves others unchanged. It properly uses .join() to build the final string. However, the prompt says "lowercase letters only," but the code applies the mapping to ANY character in the dictionary, not specifically checking if it's a lowercase letter. If the mapping contains uppercase keys, they would also be translated. This is a minor spec violation.
# Complexity:
O(n) time where n is the length of text (single pass). O(n) space for the result_chars list. This is optimal. Using list + join is O(n), whereas repeated string concatenation would be O(n²) due to string immutability.
# Robustness: 
Works for empty strings and strings without any mappable characters. However, doesn't validate input types. The function doesn't explicitly enforce "lowercase letters only" as stated in the prompt.
# Readability:
Clear variable names and helpful comments. The logic is straightforward. However, the comment about string immutability is missing—the prompt asked to "explain why strings are immutable." No docstring present. Emoji in test comments is non-standard.
# Faithfulness: 
Correctly demonstrates the list-join pattern for efficient string building, which is a key concept from Lesson One - Strings. The approach avoids the O(n²) cost of repeated concatenation due to string immutability. However, it doesn't explicitly explain the immutability concept as requested.

# Reference: Lesson One - Strings (string immutability, efficient string building with list + .join(), iteration)

# Improved Code

In [5]:
def translate(text: str, mapping: dict[str, str]) -> str:
    """
    Apply a character mapping to text, translating only lowercase letters.
    
    Strings in Python are immutable—each modification creates a new string
    object. Repeated concatenation (result += char) would be O(n²) because
    each += creates a new string and copies all previous characters.
    
    Instead, we use the list-join pattern: build a list of characters in O(n),
    then join once at the end, achieving O(n) total time.
    
    Args:
        text: The input string to translate
        mapping: Dictionary mapping lowercase letters to replacement strings
        
    Returns:
        Translated string with mappings applied to lowercase letters only
    """
    result_chars = []
    for ch in text:
        # Only apply mapping if character is a lowercase letter in the mapping
        if ch.islower() and ch in mapping:
            result_chars.append(mapping[ch])
        else:
            result_chars.append(ch)
    return ''.join(result_chars)


if __name__ == "__main__":
    leet_map = {'a': '4', 'e': '3', 'i': '1', 'o': '0'}
    
    # Basic functionality
    assert translate("Hello, World!", leet_map) == "H3ll0, W0rld!"
    assert translate("Python is awesome", leet_map) == "Pyth0n 1s 4w3s0m3"
    
    # Edge cases
    assert translate("", leet_map) == ""
    assert translate("XYZ123", leet_map) == "XYZ123"  # No lowercase mappable chars
    assert translate("aaa", leet_map) == "444"
    assert translate("AEI", leet_map) == "AEI"  # Uppercase not translated
    
    print("All tests passed.")

All tests passed.


# Task 3. Currency Formatter & Rounding Ties-to-Even. 
Write fmt_money(x: float)
-> str that returns a string like 1,234.50 using proper rounding (banker’s rounding / ties-to-even). Include test cases that demonstrate ties (e.g., 2.5, 3.5). Explain
2
floating-point representation error and how formatting mitigates it for display. Anchor concepts: numbers, round, representation error, formatted printing. Reference:
:contentReference[oaicite:8]index=8

# Prompt (verbatim)
Write a Python function fmt_money(x) that returns a string formatted like 1,234.50, using proper rounding (banker’s rounding or ties-to-even). Include test cases that demonstrate rounding of ties (e.g., 2.5, 3.5). Explain floating-point representation error and how formatted printing helps mitigate it for display.

# ChatGPT Code

In [6]:
def fmt_money(x):
    # Use round(x, 2) with banker's rounding (ties-to-even)
    rounded = round(x, 2)

    # Format with comma separators and 2 decimal places
    return f"{rounded:,.2f}"


# 🔍 Quick tests
print(fmt_money(1234.5))      # → "1,234.50"
print(fmt_money(1234.5678))   # → "1,234.57"
print(fmt_money(2.5))         # → "2.00"  (2 is even → rounds down)
print(fmt_money(3.5))         # → "4.00"  (3 is odd → rounds up)
print(fmt_money(1234567.345)) # → "1,234,567.34"


1,234.50
1,234.57
2.50
3.50
1,234,567.34


# Critique

# Correctness: 
The function correctly uses banker's rounding via round(x, 2) and formats with comma separators and 2 decimal places using f-string format specifiers. However, the test case comments have errors: 2.5 rounds to 2.50 (not 2.00), and 3.5 rounds to 4.00 correctly. The logic is sound but test expectations are misleading.
# Complexity: 
O(1) time for the rounding and formatting operations. O(log n) space where n is the magnitude of the number (for the string representation). This is optimal.
# Robustness: 
Works for positive numbers but doesn't explicitly handle edge cases like negative numbers, zero, very large numbers, or non-numeric input. No type checking present.
# Readability: 
Clear variable names and comments explain the approach. However, the prompt requested an explanation of floating-point representation error and how formatted printing mitigates it—this explanation is completely missing from the code. The intermediate rounded variable is unnecessary; the format specifier :,.2f already handles rounding. Emoji in comments is non-standard.
# Faithfulness: 
Uses f-string formatting with format specifiers (:,.2f), which is a string formatting concept from Lesson One - Strings. However, the code doesn't explain the floating-point issue as requested. The explicit round() call is redundant since the .2f format specifier already rounds to 2 decimal places.

# Reference: Lesson One - Strings (f-strings, format specifiers)

# Improved Code

In [14]:
def fmt_money(x: float) -> str:
    """
    Format a number as money with comma separators and 2 decimal places.
    
    Floating-point representation error: Binary floats cannot exactly represent
    many decimal values (e.g., 0.1 + 0.2 = 0.30000000000000004).
    
    The format specifier `:,.2f` mitigates this by rounding to 2 decimal places
    and formatting as a fixed-point string with comma separators.
    """
    return f"{x:,.2f}"


print(fmt_money(1234.5))
print(fmt_money(1234.5678))
print(fmt_money(2.126))
print(fmt_money(2.124))
print(fmt_money(2.5))
print(fmt_money(3.5))
print(fmt_money(2.125))
print(fmt_money(1234567.89))
print(fmt_money(9876543.21))
print(fmt_money(0))
print(fmt_money(0.99))
print(fmt_money(-1234.5))
print(fmt_money(0.1 + 0.2))

1,234.50
1,234.57
2.13
2.12
2.50
3.50
2.12
1,234,567.89
9,876,543.21
0.00
0.99
-1,234.50
0.30


# Task 4. Exponent Tool & Input Validation. 
Create a CLI program exponent.py that
reads two values, validates they are numeric, handles errors gracefully (e.g., ValueError),
and prints a ** b with clear messages. Compare using try/except vs. pre-validation
with str.isdigit() (and why isdigit() is insufficient for negatives/floats). Anchor concepts: arithmetic operators, exceptions, input. Reference: :contentReference[oaicite:9]index=9, :contentReference[oaicite:10]index=10


# Prompt (paste verbatim)
Create a Python CLI program exponent.py that reads two values, validates they are numeric, handles errors gracefully (e.g., ValueError), and prints a ** b with clear messages. Compare using try/except versus pre-validation with str.isdigit(), and explain why isdigit() is insufficient for negatives or floats. Add simple test runs or input/output examples.

# ChatGPT Code

In [25]:

a = input("Enter base: ")
b = input("Enter exponent: ")

if a.isdigit() and b.isdigit():
    print(f"Result: {float(a)} ** {float(b)} = {float(a) ** float(b)}\n")
else:
    print("Invalid input. Please enter numbers only.")

Enter base:  2
Enter exponent:  3


Result: 2.0 ** 3.0 = 8.0



# Critique

# Correctness: 
The program works for positive integers and prints correct results.

    Lesson References:

        Uses input() and ** (exponentiation) from Lesson Two – Numbers & Math.

        The if/else decision structure connects to Lesson Four – Conditionals.

# Limitations: 
.isdigit() cannot handle negative (-3) or float (2.5) numbers since these include non-digit characters (-, .).

# Error Handling:
Basic validation only; lacks robust handling of unexpected input.

# Faithfulness: 
Implements early lesson concepts correctly, but doesn’t yet use functions or exception handling from later lessons.

# Improvement 
Opportunity: Could apply functions (Lesson Three), try/except (Lesson Four), and modules (Lesson Seven) for better structure and maintainability.

# Improved Code

In [26]:
def main():
    try:
        """Read two numbers, compute base ** exponent with safe conversion."""
        base_str = input("Enter base: ")
        exp_str = input("Enter exponent: ")
        
        base = float(base_str)
        exponent = float(exp_str)
        
        result = base ** exponent
        print(f"Result: {base} ** {exponent} = {result}\n")
        
    except ValueError:
        print("Error: inputs must be numeric (e.g., 3, -2.5, 1e3).")

if __name__ == "__main__":
    main()

Enter base:  2
Enter exponent:  3


Result: 2.0 ** 3.0 = 8.0



# Task 5 Password Policy Checker (If/Elif/Else + Loops). 
Implement check_password(pw:
str) -> dict that validates: length ≥ 10, at least one uppercase, one lowercase,
one digit, one symbol in !@#$%^&*, and no spaces. Return a dict of booleans per
rule and an overall pass/fail. Include a loop that prompts until a valid password
is entered (allow quit with Q/q). Anchor concepts: conditionals, logical operators,
loops, break/continue. Reference: :contentReference[oaicite:11]index=11

# Prompt (paste verbatim)
Write a Python function check_password(pw) that validates a password according to the following rules: it must be at least 10 characters long, contain at least one uppercase letter, one lowercase letter, one digit, and one symbol from !@#$%^&*, and it must not contain any spaces. The function should return a dictionary of booleans for each rule as well as an overall pass/fail result. Additionally, include a loop that repeatedly prompts the user to enter a password until it passes all checks, allowing the user to quit by entering Q or q. Use if/elif/else statements, logical operators, and loop control statements such as break or continue.

# ChatGPT Code

In [27]:
while True:
    pw = input("Enter password (Q to quit): ")

    if pw.lower() == "q":
        print("Goodbye!")
        break

    has_upper = False
    has_lower = False
    has_digit = False
    has_symbol = False
    has_space = False

    for ch in pw:
        if ch.isupper():
            has_upper = True
        elif ch.islower():
            has_lower = True
        elif ch.isdigit():
            has_digit = True
        elif ch in "!@#$%^&*":
            has_symbol = True
        elif ch.isspace():
            has_space = True

    if len(pw) >= 10 and has_upper and has_lower and has_digit and has_symbol and not has_space:
        print("Password is valid!\n")
        break
    else:
        print("Invalid password. Please try again.\n")


Enter password (Q to quit):  123123123131


Invalid password. Please try again.



Enter password (Q to quit):  Gwapo0123


Invalid password. Please try again.



Enter password (Q to quit):  @Gwapo09123


Password is valid!



# Critique

# Correctness: 
The program correctly checks for uppercase, lowercase, digits, symbols, and spaces, and it identifies valid and invalid passwords for most typical cases.

     Lesson References: 
        The code demonstrates the use of if/elif/else and logical operators from Lesson Four – Conditionals, as well as loops from Lesson Three – Functions and Loops, and   applies basic string methods introduced in Lesson One – Strings.

# Limitations: 
The program is limited because it does not encapsulate the validation logic into a function, and it does not return a structured dictionary of rule checks.

# Error Handling: 
While the program allows the user to quit with 'Q/q', it lacks a structured approach to report which specific rules failed, providing only a generic invalid message.

# Faithfulness: 
This implementation is faithful to the concepts taught in the early lessons but does not yet incorporate functions, modular design, or more advanced handling from higher lessons.

# Improvement Opportunity: 
The program could be improved by creating a check_password() function that returns a dictionary of validation results, using loops and logical operators more efficiently, and organizing the code with a main() function for better readability and maintainability, in line with lessons three through seven.

# Improved Code

In [29]:
def check_password(pw: str) -> dict:
    """Validate password according to policy rules."""
    results = {
        "length": len(pw) >= 10,
        "uppercase": any(c.isupper() for c in pw),
        "lowercase": any(c.islower() for c in pw),
        "digit": any(c.isdigit() for c in pw),
        "symbol": any(c in "!@#$%^&*" for c in pw),
        "no_space": not any(c.isspace() for c in pw)
    }
    results["passed"] = all(results.values())
    return results


def main():
    """Loop until a valid password is entered or user quits."""
    while True:
        pw = input("Enter password (Q to quit): ")
        if pw.lower() == "q":
            print("Goodbye!")
            break

        result = check_password(pw)
        if result["passed"]:
            print("Password is valid!\n")
            break
        else:
            print("Password failed these rules:")
            for rule, ok in result.items():
                if rule != "passed" and not ok:
                    print(f" - {rule}")
            print("Please try again.\n")


if __name__ == "__main__":
    main()


Enter password (Q to quit):  gwapo09222!


Password failed these rules:
 - uppercase
Please try again.



Enter password (Q to quit):  wapo09


Password failed these rules:
 - length
 - uppercase
 - symbol
Please try again.



Enter password (Q to quit):  @gwapo2300Ayy


Password is valid!



#  Task 6. Dice Simulation (Monte Carlo). 
Write roll() -> int to simulate a fair die,
then simulate 100,000 rolls to estimate the probability that the sum of two dice
equals 7. Compare empirical result to the exact probability. Comment on random
seeding and reproducibility. Anchor concepts: random module, simulation, averages.
Reference: :contentReference[oaicite:12]index=12


 # Prompt (paste verbatim)
Write a Python function roll() that simulates a fair six-sided die and returns an integer between 1 and 6. Then simulate 100,000 rolls of two dice to estimate the probability that their sum equals 7. Compare your empirical result to the exact probability and comment on the effect of random seeding and reproducibility. Use the random module and discuss simulation and averaging techniques.

# ChatGPT Code

In [30]:
import random

def roll():
    """Simulate rolling a fair 6-sided die."""
    return random.randint(1, 6)

# Simulate two dice rolls 100,000 times
count_sum_7 = 0
for _ in range(100000):
    if roll() + roll() == 7:
        count_sum_7 += 1

probability = count_sum_7 / 100000
print(f"Estimated probability that sum equals 7: {probability}")


Estimated probability that sum equals 7: 0.16636


# Critique

# Correctness: 
The program correctly simulates rolling two dice 100,000 times and counts how often the sum equals 7, producing a reasonable empirical probability close to the exact probability of 1/6.

    Lesson References: This code demonstrates the use of the random module from Lesson Seven – Packages to simulate randomness, and uses loops and basic arithmetic from Lesson Two – Numbers & Math to calculate averages and probability.

# Limitations: 
The simulation is simple but not reproducible because it does not set a random seed. Results will differ on each run, which could be important for debugging or testing.

# Error Handling: 
No input is required from the user, so the program does not currently handle any errors. However, potential improvements could include validating the number of simulations if it were user-supplied.

# Faithfulness: 
The program is faithful to Monte Carlo simulation concepts but does not yet use higher-level structure like a main() function or separate functions for rolling two dice or computing probabilities.

# Improvement Opportunity: 
The program could be improved by defining a main() function, adding a separate function to roll two dice, allowing the number of simulations to be input by the user, and using random.seed() for reproducibility. This would incorporate lessons on functions (Lesson Three) and modularity (Lesson Seven).

# Improved Code

In [32]:
import random

def roll() -> int:
    """Simulate rolling a fair 6-sided die."""
    return random.randint(1, 6)

def simulate_rolls(n: int, target_sum: int) -> float:
    """Simulate n rolls of two dice and estimate probability of target_sum."""
    count = 0
    for _ in range(n):
        if roll() + roll() == target_sum:
            count += 1
    return count / n

def main():
    """Run Monte Carlo simulation for sum of two dice equals 7."""
    random.seed(42) 
    num_simulations = 100000
    target_sum = 7
    prob = simulate_rolls(num_simulations, target_sum)
    exact_prob = 6 / 36 
    print(f"Estimated probability: {prob}")
    print(f"Exact probability: {exact_prob}")

if __name__ == "__main__":
    main()


Estimated probability: 0.16456
Exact probability: 0.16666666666666666


# Task 7. OOP: Vehicles with Odometer and Fuel. Design classes Vehicle (base),
Car and Truck (children). Each has color, mileage, and fuel_liters. Provide
drive(km, km_per_liter) which increments mileage and decreases fuel (never below 0) and refuel(liters). Show how inheritance avoids duplication and add
a method override in Truck (e.g., different efficiency). Include a minimal test
script. Anchor concepts: classes vs. instances, attributes, instance methods, inheritance/override. Reference: :contentReference[oaicite:13]index=13


# Prompt (paste verbatim)
Design Python classes for a vehicle system with a base class Vehicle and child classes Car and Truck. Each class should have attributes for color, mileage, and fuel in liters. Implement a drive(km, km_per_liter) method that increments mileage and decreases fuel without letting it drop below zero, and a refuel(liters) method to add fuel. Show how inheritance avoids code duplication and include a method override in Truck (for example, with different fuel efficiency). Provide a minimal test script demonstrating these features, using classes versus instances, instance attributes and methods, and inheritance/overriding.

# ChatGPT Code

In [33]:
class Vehicle:
    def __init__(self, color, mileage, fuel_liters):
        self.color = color
        self.mileage = mileage
        self.fuel_liters = fuel_liters

    def drive(self, km, km_per_liter):
        needed_fuel = km / km_per_liter
        if needed_fuel > self.fuel_liters:
            km = self.fuel_liters * km_per_liter
            self.fuel_liters = 0
        else:
            self.fuel_liters -= needed_fuel
        self.mileage += km

    def refuel(self, liters):
        self.fuel_liters += liters


class Car(Vehicle):
    pass


class Truck(Vehicle):
    def drive(self, km, km_per_liter):
        # Truck uses 80% efficiency
        km_per_liter *= 0.8
        super().drive(km, km_per_liter)


# Minimal test
car = Car("red", 10000, 50)
truck = Truck("blue", 20000, 80)

car.drive(100, 10)
truck.drive(100, 10)

print(f"Car mileage: {car.mileage}, fuel left: {car.fuel_liters}")
print(f"Truck mileage: {truck.mileage}, fuel left: {truck.fuel_liters}")


Car mileage: 10100, fuel left: 40.0
Truck mileage: 20100, fuel left: 67.5


# Critique

# Correctness: 
The program correctly models vehicles, updates mileage and fuel when driving, and refueling works as expected. Trucks have a modified drive method demonstrating method override.

    Lesson References: 
        This code illustrates classes, instance attributes, and instance methods from Lesson Six – OOP. Inheritance is shown by Car and Truck deriving from Vehicle, and Truck overrides drive() to apply different efficiency.

# Limitations: 
The Car class does not modify behavior from Vehicle, which is fine, but edge cases such as negative fuel or negative drive distances are not checked.

# Error Handling: 
The code does not include type or value checks; it assumes correct numeric inputs for distances and fuel.

# Faithfulness: 
This implementation demonstrates object-oriented principles faithfully, showing class definition, inheritance, instance creation, and method overriding.

# Improvement Opportunity:
The code could be improved by adding input validation, encapsulating the test script in a main() function, and possibly extending with more vehicle types or more realistic fuel efficiency rules, connecting to higher lessons on modularity and structured testing.

# Improved Code

In [34]:
class Vehicle:
    def __init__(self, color: str, mileage: float, fuel_liters: float):
        self.color = color
        self.mileage = mileage
        self.fuel_liters = fuel_liters

    def drive(self, km: float, km_per_liter: float):
        """Drive the vehicle, decreasing fuel and increasing mileage."""
        needed_fuel = km / km_per_liter
        if needed_fuel > self.fuel_liters:
            km = self.fuel_liters * km_per_liter
            self.fuel_liters = 0
        else:
            self.fuel_liters -= needed_fuel
        self.mileage += km

    def refuel(self, liters: float):
        """Refuel the vehicle."""
        self.fuel_liters += liters


class Car(Vehicle):
    pass


class Truck(Vehicle):
    def drive(self, km: float, km_per_liter: float):
        """Override drive: truck is 80% as fuel-efficient."""
        km_per_liter *= 0.8
        super().drive(km, km_per_liter)


def main():
    """Test script for vehicles."""
    car = Car("red", 10000, 50)
    truck = Truck("blue", 20000, 80)

    car.drive(100, 10)
    truck.drive(100, 10)

    print(f"Car mileage: {car.mileage}, fuel left: {car.fuel_liters}")
    print(f"Truck mileage: {truck.mileage}, fuel left: {truck.fuel_liters}")


if __name__ == "__main__":
    main()


Car mileage: 10100, fuel left: 40.0
Truck mileage: 20100, fuel left: 67.5


# Task 8. OOP + Composition: Kennel with Dogs. 
Using a Dog class (name, age, breed,
bark()), implement a Kennel manager that holds multiple Dog instances, supports
add_dog, remove_dog, oldest_dog(), and find_by_breed(breed). Compare composition vs. inheritance for this scenario. Anchor concepts: OOP design, managing
collections of objects. Reference: :contentReference[oaicite:14]index=14


# Prompt (paste verbatim)
Using a Dog class with attributes name, age, breed, and a bark() method, implement a Kennel manager that holds multiple Dog instances. The Kennel should support adding and removing dogs, finding the oldest dog, and searching for dogs by breed using find_by_breed(breed). Discuss how composition is used in this scenario and compare it to inheritance. Focus on OOP design and managing collections of objects.

# ChatGPT Code

In [43]:
class Dog:
    def __init__(self, name, age, breed):
        self.name = name
        self.age = age
        self.breed = breed

    def bark(self):
        print(f"{self.name} says: Woof!")


class Kennel:
    def __init__(self):
        self.dogs = []

    def add_dog(self, dog):
        self.dogs.append(dog)

    def remove_dog(self, dog_name):
        self.dogs = [dog for dog in self.dogs if dog.name != dog_name]

    def oldest_dog(self):
        if not self.dogs:
            return None
        return max(self.dogs, key=lambda d: d.age)

    def find_by_breed(self, breed):
        return [dog for dog in self.dogs if dog.breed == breed]


# Minimal test
dog1 = Dog("Rex", 5, "Labrador")
dog2 = Dog("Bella", 7, "Beagle")
dog3 = Dog("Max", 3, "Labrador")

kennel = Kennel()
kennel.add_dog(dog1)
kennel.add_dog(dog2)
kennel.add_dog(dog3)

oldest = kennel.oldest_dog()
print(f"Oldest dog: {oldest.name}, age {oldest.age}")

labradors = kennel.find_by_breed("Labrador")
print(f"Labradors: {[dog.name for dog in labradors]}")


Oldest dog: Bella, age 7
Labradors: ['Rex', 'Max']


# Critique

# Correctness: 
The program correctly manages multiple Dog instances, supports adding and removing dogs, finds the oldest dog, and can filter dogs by breed.

    Lesson References: 
        This code demonstrates object-oriented design principles from Lesson Six – OOP, including class definitions, instance attributes, instance methods, and managing collections of objects using lists.

# Limitations: 
The code does not handle duplicate names when removing dogs, and oldest_dog() returns None for an empty kennel without additional user feedback.

# Error Handling:
No input validation is included, so it assumes proper Dog instances are passed to the Kennel methods.

# Faithfulness: 
The program illustrates composition well, showing that a Kennel contains Dog instances rather than inheriting from Dog. This makes sense because a kennel “has dogs” instead of “is a dog.”

# Improvement Opportunity: 
The program could be improved by adding input validation, encapsulating tests in a main() function, and adding methods to handle edge cases like duplicate names or empty kennels. This would connect to higher lessons on functions (Lesson Three), modularity (Lesson Seven), and managing collections (Lesson Five).

# Improved Code

In [42]:
class Dog:
    def __init__(self, name: str, age: int, breed: str):
        self.name = name
        self.age = age
        self.breed = breed

    def bark(self):
        print(f"{self.name} says: Woof!")


class Kennel:
    def __init__(self):
        self.dogs = []

    def add_dog(self, dog: Dog):
        self.dogs.append(dog)

    def remove_dog(self, dog_name: str):
        self.dogs = [dog for dog in self.dogs if dog.name != dog_name]

    def oldest_dog(self):
        if not self.dogs:
            return None
        return max(self.dogs, key=lambda d: d.age)

    def breeds(self):
        """Return a set of all breeds in the kennel."""
        return set(dog.breed for dog in self.dogs)

    def find_by_breed(self, breed: str):
        return [dog for dog in self.dogs if dog.breed == breed]


def main():
    """Test script for the Kennel and Dog classes."""
    dog1 = Dog("Rex", 5, "Labrador")
    dog2 = Dog("Bella", 7, "Beagle")
    dog3 = Dog("Max", 3, "Labrador")
    dog4 = Dog("Lucy", 4, "Beagle")
    dog5 = Dog("Charlie", 2, "Poodle")

    kennel = Kennel()
    for dog in [dog1, dog2, dog3, dog4, dog5]:
        kennel.add_dog(dog)

    oldest = kennel.oldest_dog()
    if oldest:
        print(f"Oldest dog: {oldest.name}, age {oldest.age}")

    print("\nDogs by breed:")
    for breed in kennel.breeds():
        breed_dogs = kennel.find_by_breed(breed)
        print(f"{breed}: {[dog.name for dog in breed_dogs]}")


if __name__ == "__main__":
    main()


Oldest dog: Bella, age 7

Dogs by breed:
Poodle: ['Charlie']
Beagle: ['Bella', 'Lucy']
Labrador: ['Rex', 'Max']


# Task 9. Modules and Packages: Helpers. 
Create a package helpers/ with modules
string_utils.py (function shout(s) uppercases) and math_utils.py (function area(l,w)).
Then a main.py that imports with both import ... as ... and from ...
import .... Explain namespace collisions and why aliases help. Anchor concepts: modules, packages, import variations, namespaces. Reference: :contentReference[oaicite:15]index=15

# Prompt (paste verbatim)
Create a Python package helpers/ containing the modules string_utils.py with a function shout(s) that returns the uppercase version of a string, and math_utils.py with a function area(l, w) that computes the area of a rectangle. Then create a main.py that imports these modules using both import ... as ... and from ... import ... syntax. Explain namespace collisions and why using aliases can help. Focus on modules, packages, import variations, and namespaces.

# ChatGPT Code

In [49]:
# Import modules using alias to avoid potential namespace collisions
# Importing the whole module with alias
from helpers.string_utils import shout

# Importing a specific function directly
from helpers.math_utils import area

text = "hello world"
print(shout(text))

length = 5
width = 3
print(area(length, width))


HELLO WORLD
15


# Critique

# Correctness: 
The program correctly implements a package with two modules and imports them in main.py. The shout function converts text to uppercase, and area calculates the area of a rectangle.

    Lesson References: 
        This code demonstrates the use of packages and modules from Lesson Seven – Packages, shows both import ... as ... and from ... import ... variations, and illustrates the concept of namespaces.

# Limitations: 
The code does not handle incorrect inputs, such as non-string input to shout or non-numeric input to area.

# Error Handling: 
There is no error handling, so the functions assume correct input types.

# Faithfulness: 
The code faithfully demonstrates modules, package structure, and import aliases, which helps avoid namespace collisions when multiple modules have functions with the same name.

# Improvement Opportunity: 
The program could be improved by adding type hints, docstrings, and possibly demonstrating a scenario where namespace collisions occur to emphasize why aliases and from ... import ... are useful.

# Improved Code

In [51]:
# Importing a specific function directly
from helpers.string_utils import shout_improved as shout

# Importing a specific function directly
from helpers.math_utils import area_improved as area_rec
def main():
    """Demonstrate usage of helper modules."""
    text = "hello world"
    print(shout(text))

    length = 5
    width = 3
    print(f"Rectangle area: {area_rec(length, width)}")

if __name__ == "__main__":
    main()

HELLO WORLD
Rectangle area: 15


# Task 10.  Robust Temperature Converter (Functions + Exceptions). 
Write two functions: cel_to_far(c) and far_to_cel(f) with type hints and docstrings. Build a
CLI that asks the user which direction to convert and validates input (gracefully
handles bad entries). Include unit tests (doctest or simple asserts) covering typical
3
and edge cases (e.g., -40). Anchor concepts: writing functions, control flow, input
handling, testing.


# Prompt (paste verbatim)
Write two Python functions, cel_to_far(c) and far_to_cel(f), with type hints and docstrings that convert temperatures between Celsius and Fahrenheit. Build a CLI that asks the user which direction to convert and validates input, handling bad entries gracefully. Include unit tests using either doctest or simple assert statements that cover typical and edge cases (for example, -40). Focus on writing functions, control flow, input handling, and testing.

# ChatGPT Code

In [55]:
def cel_to_far(c):
    return c * 9/5 + 32

def far_to_cel(f):
    return (f - 32) * 5/9

direction = input("Convert to Celsius or Fahrenheit? (C/F): ").strip().upper()

if direction == "C":
    f = float(input("Enter Fahrenheit: "))
    print(f"{f}°F is {far_to_cel(f):.2f}°C")
elif direction == "F":
    c = float(input("Enter Celsius: "))
    print(f"{c}°C is {cel_to_far(c):.2f}°F")
else:
    print("Invalid choice")


Convert to Celsius or Fahrenheit? (C/F):  c
Enter Fahrenheit:  32


32.0°F is 0.00°C


# Critique

# Correctness: 
The program correctly converts temperatures in both directions and prints the results formatted to two decimal places.

    Lesson References: 
        This code demonstrates function definitions from Lesson Three – Functions and Loops, input handling with input() and float() conversions, and conditional control flow from Lesson Four – Conditionals.

# Limitations: 
The program does not handle invalid numeric inputs, so entering text instead of a number would raise a ValueError. Edge cases, like negative temperatures or very large values, are not explicitly tested.

# Error Handling: 
Only invalid direction choices are handled; the numeric inputs are assumed valid, which could cause runtime errors.

# Faithfulness: 
The implementation uses functions and basic control flow correctly but lacks modularity, type hints, docstrings, and automated testing that were requested in the prompt.

# Improvement Opportunity: 
The code could be improved by adding type hints, docstrings, exception handling for numeric input, a main() function for CLI structure, and simple tests using assert to cover typical and edge cases.

# Improved Code

In [56]:
def cel_to_far(c: float) -> float:
    """Convert Celsius to Fahrenheit."""
    return c * 9 / 5 + 32

def far_to_cel(f: float) -> float:
    """Convert Fahrenheit to Celsius."""
    return (f - 32) * 5 / 9

def main():
    """CLI for converting temperatures between Celsius and Fahrenheit."""
    while True:
        direction = input("Convert to Celsius or Fahrenheit? (C/F, Q to quit): ").strip().upper()
        if direction == "Q":
            print("Goodbye!")
            break

        try:
            if direction == "C":
                f = float(input("Enter Fahrenheit: "))
                print(f"{f}°F is {far_to_cel(f):.2f}°C\n")
            elif direction == "F":
                c = float(input("Enter Celsius: "))
                print(f"{c}°C is {cel_to_far(c):.2f}°F\n")
            else:
                print("Invalid choice. Please enter C, F, or Q.\n")
        except ValueError:
            print("Invalid number. Please enter a valid numeric value.\n")


if __name__ == "__main__":
    main()

# Simple tests
assert cel_to_far(0) == 32
assert cel_to_far(-40) == -40
assert far_to_cel(32) == 0
assert far_to_cel(-40) == -40
assert round(cel_to_far(100), 2) == 212.00
assert round(far_to_cel(212), 2) == 100.00


Convert to Celsius or Fahrenheit? (C/F, Q to quit):  f
Enter Celsius:  213.21


213.21°C is 415.78°F



Convert to Celsius or Fahrenheit? (C/F, Q to quit):  q


Goodbye!
